core/src/main/java/org/infinispan/stream/impl/AbstractCacheStream.java

```diff
@@ -762,73 +711,30 @@ public void accept(Iterable<E> es) {
       }
    }
 
-   enum IntermediateType {
-      OBJ,
-      INT,
-      DOUBLE,
-      LONG,
-      NONE {
-         @Override
-         public boolean shouldUseIntermediate(boolean sorted, boolean distinct) {
-            return false;
+   /**
+    * Given two SegmentCompletionListener, return a SegmentCompletionListener that
+    * executes both in sequence, even if the first throws an exception, and if both
+    * throw exceptions, add any exceptions thrown by the second as suppressed
+    * exceptions of the first.
+    */
+   protected static CacheStream.SegmentCompletionListener composeWithExceptions(CacheStream.SegmentCompletionListener a,
+           CacheStream.SegmentCompletionListener b) {
+      return (segments) -> {
+         try {
+            a.segmentCompleted(segments);
+         }
+         catch (Throwable e1) {
+            try {
+               b.segmentCompleted(segments);
+            }
+            catch (Throwable e2) {
+               try {
+                  e1.addSuppressed(e2);
+               } catch (Throwable ignore) {}
```

Comment:
If there's an exception happening in `addSuppressed()`, I think it's important enough to throw that instead of the original exception.